payments-ui-core/src/main/java/com/stripe/android/ui/core/elements/CardNumberController.kt

```diff
@@ -78,4 +118,55 @@ internal class CardNumberController constructor(
     override fun onFocusChange(newHasFocus: Boolean) {
         _hasFocus.value = newHasFocus
     }
+
+    private fun shouldQueryRepository(
+        accountRange: AccountRange
+    ) = when (accountRange.brand) {
+        CardBrand.Unknown,
+        CardBrand.UnionPay -> true
+        else -> false
+    }
+
+    @JvmSynthetic
+    internal fun queryAccountRangeRepository(cardNumber: CardNumber.Unvalidated) {
+        if (shouldQueryAccountRange(cardNumber)) {
+            // cancel in-flight job
+            cancelAccountRangeRepositoryJob()
+
+            // invalidate accountRange before fetching
+            accountRange = null
+
+            accountRangeRepositoryJob = CoroutineScope(Dispatchers.IO).launch {
+                val bin = cardNumber.bin
+                val accountRange = if (bin != null) {
+                    cardAccountRangeRepository.getAccountRange(cardNumber)
+                } else {
+                    null
+                }
+
+                withContext(Dispatchers.Main) {
+                    onAccountRangeResult(accountRange)
+                }
+            }
+        }
+    }
+
+    private fun cancelAccountRangeRepositoryJob() {
+        accountRangeRepositoryJob?.cancel()
+        accountRangeRepositoryJob = null
+    }
+
+    @JvmSynthetic
+    internal fun onAccountRangeResult(
+        newAccountRange: AccountRange?
+    ) {
+        accountRange = newAccountRange
+        (visualTransformation as CardNumberVisualTransformation).maxPanLength = 19
+    }
+
+    private fun shouldQueryAccountRange(cardNumber: CardNumber.Unvalidated): Boolean {
+        return accountRange == null ||
+            cardNumber.bin == null ||
+            accountRange?.binRange?.matches(cardNumber) == false
+    }
```

Comment:
I wonder how much of this logic is also in the CardNumberEditText?  Is there a wrapper class or additional functions we should add to the existing classes to reuse this logic?

Reply:
Most of this highlighted section is shared, except the visualTransformation part; we can't make a superclass since CardNumberEditText already inherits from a superclass, but making a shared interface would be possible

Reply:
I think we should circle back to this in a different PR. I think having a shared interface to give us a level of abstraction would help.

 I'll sync with @epan-stripe in the next week or so to find some time to slot this in.